### PR TITLE
[RPC] Expose ancestor/descendant information over RPC

### PR DESCRIPTION
**Pull request by sdaftuar**
This adds 3 new RPC calls (~~getancestors~~ getmempoolancestors, ~~getdescendants~~ getmempooldescendants, getmempoolentry) to expose more mempool information over RPC.

Now that we have policy rules that are tied to transaction chains (including limits on the number of in-mempool ancestors and in-mempool descendants, and mempool eviction and RBF policies that depend on fees of descendants), it seems helpful to be able to expose more information about the chains a given tx is part of over RPC.

This is motivated by the discussion in #7222.  @petertodd You mentioned wanting to see specific use-cases for this information; I think there are clear improvements to the fee-bumping tools in your replace-by-fee repo that could be made if we had these functions (the first two tools I looked at just now don't seem to consider transaction chains at all in fee calculations?).  I'll try to post a link in this PR to specific proposed improvements when I have something to share that will demonstrate this more clearly.

### CODE REVIEW

**Comment by btcdrak** [comment]
Concept ACK

**Comment by jonasschnelli** [comment]
Concept ACK.
Some concrete use-cases would be nice.
Nice to see the new RPC commands `cs_main` lock free.

**Comment by sdaftuar** [comment]
Updated to remove dead code I inadvertently left in `rpcblockchain.cpp`.

**Comment by sdaftuar** [comment]
Here's an example of a tool that would benefit from the getdescendants function:
https://github.com/sdaftuar/replace-by-fee-tools/blob/d0121b03d02c1a7e5e16425b55da1302ddf081aa/combine-transactions.py

This is a script that takes two txid's (that are signaling opt-in RBF) and combines them down to one -- consolidating duplicate output addresses (if any) and removing pay-to-self outputs, assumed for demo purposes to be change outputs -- and paying for all fees including those of descendants.  The main idea is that this avoids over-paying for descendants -- it calls getdescendants on each transaction and takes the union of the two sets so that each descendant is paid for once.  I don't know any simple or straightforward way to do this without this new RPC call.

There are similar benefits to having getancestors, such as if you wanted to understand why a stuck transaction wasn't confirming, then getancestors would give you the set of unconfirmed parents to look at (and you could write a replace-by-fee-tool that was smart enough to drop an input coming from a stuck unconfirmed parent, when possible, to replace a transaction with one likely to confirm sooner; or better still, replace such a stuck parent instead if possible).

**Comment by sipa** [comment]
Ultra nit before reading the code: it's not obvious from the RPC names that
it's about the mempool.

**Comment by sdaftuar** [comment]
@sipa True.  Perhaps `getmempoolancestors/getmempooldescendants`?  I'm open to suggestions.

@laanwj if this code is ok I think it would make sense to include this in 0.12

**Comment by sdaftuar** [comment]
Updated with new RPC names (`getmempoolancestors` and `getmempooldescendants`).

**Comment by luke-jr** [comment]
Any reason not to just add the ancestors/descendants info to getrawmempool (and also the getmempoolentry)?

**Comment by sdaftuar** [comment]
The goal was to avoid having to dump the entire mempool if you were only interested in a very small subset of it.  But I suppose there's no reason we couldn't also show the txid's of ancestors and descendants when you call `getrawmempool` (with verbose=true).  If that seems useful, I'm happy to add that.

`getmempoolentry` should be returning exactly the same information for a single transaction that `getrawmempool` shows for all transactions.

**Comment by sdaftuar** [comment]
@luke-jr  On further thought, the downside to that is that doing the calculation could cause `getrawmempool` to get up to ~25x slower (default policy allows length 25 ancestor and descendant chains, and the calculation of ancestors and descendants requires walking those chains), which on a full mempool could be a meaningful -- and perhaps unacceptable? -- slowdown.

The other possibility would be to add a new option to `getrawmempool` to determine whether to do the calculation or not, but I think we try to avoid adding new arguments to existing RPC calls?  So now I think it's probably better to leave as-is.  Let me know if you disagree...

**Comment by laanwj** [comment]
Concept ACK

**Comment by laanwj** [comment]
Needs rebase.

**Comment by sdaftuar** [comment]
Rebased.  There weren't any actual conflicts though, so not sure why github thought it wouldn't merge cleanly?

**Comment by laanwj** [comment]
Thanks!
I've noticed that before. Github is somewhat less good at merging than git.

**Comment by sipa** [comment]
Rebase needed.

**Comment by sdaftuar** [comment]
Rebased.  Also added a commit that extends the mempool entry output to include the ancestor state (added by #7594), as well as a brief update to the release notes.

File: src/rpc/blockchain.cpp [hunk 1]
```diff
@@ -280,6 +298,167 @@ UniValue getrawmempool(const UniValue& params, bool fHelp)
...
+    if (fHelp || params.size() < 1 || params.size() > 2) {
+        throw runtime_error(
+            "getmempooldescendants txid (verbose)\n"
+            "\nIf txid is in the mempool, returns all in-mempool descendants.\n"
+            "\nArguments:\n"
+            "1. \"txid\"                   (string, required) The transaction id (must be in mempool)\n"
+            "2. verbose                  (boolean, optional, default=false) true for a json object, false for array of transaction ids\n"
+            "\nResult (for verbose=false):\n"
+            "[                       (json array of string)\n"
+            "  \"transactionid\"       (string) The transaction id of an in-mempool descendant transaction\n"
```
Comment (jonasschnelli) [inline src/rpc/blockchain.cpp]: nit: missing four whitespace before `(`

**Comment by jonasschnelli** [comment]
I don't know why Travis WIN32 fails on the unit tests (sadly, It seems that the unit-test log is missing in the travis dump).

`mempool_packages.py` needs to be added to the `rpc-tests.py` script.

Tested ACK 69c9553586befdbc9ca6a474ae294bebe998163b modulo nits.

**Comment by sdaftuar** [comment]
Fixed @jonasschnelli's nits.  No idea how these changes could affect the unit tests on any platform; the travis failure seems disturbing...

File: src/rpc/blockchain.cpp [hunk 2]
```diff
@@ -292,6 +292,70 @@ UniValue getrawmempool(const UniValue& params, bool fHelp)
...
+{
+    if (fHelp || params.size() < 1 || params.size() > 2) {
+        throw runtime_error(
+            "getmempoolancestors txid (verbose)\n"
+            "\nIf txid is in the mempool, returns all in-mempool ancestors.\n"
+            "\nArguments:\n"
+            "1. \"txid\"                   (string, required) The transaction id (must be in mempool)\n"
+            "2. verbose                  (boolean, optional, default=false) true for a json object, false for array of transaction ids\n"
+            "\nResult (for verbose=false):\n"
+            "[                       (json array of string)\n"
```
Comment (paveljanik) [inline src/rpc/blockchain.cpp]: json array of string*s*

**Comment by paveljanik** [comment]
ACK https://github.com/bitcoin/bitcoin/pull/7292/commits/170bfea0260f58aaed24c230a3258cfad5b31398

**Comment by sdaftuar** [comment]
Pushed a fix to address @paveljanik's nit

**Comment by sdaftuar** [comment]
Unit tests are now failing on linux.  I'll see if I can reproduce locally.

**Comment by sdaftuar** [comment]
Oh, I suppose it could be this: https://github.com/bitcoin/bitcoin/issues/8069#issue-155434890

**Comment by sdaftuar** [comment]
Looks like the travis issue was fixed by #8070, so I think this is ready now.

**Comment by paveljanik** [comment]
reACK https://github.com/bitcoin/bitcoin/pull/7292/commits/5379307a23177c33b5a869513d1d1d09790b0269

**Comment by MarcoFalke** [comment]
utACK 5379307

**Comment by arowser** [comment]
Can one of the admins verify this patch?

**Comment by sipa** [comment]
utACK 5379307a23177c33b5a869513d1d1d09790b0269

**Comment by sdaftuar** [comment]
Is there anything holding this up?

**Comment by sipa** [comment]
Can you squash the typo?

**Comment by sdaftuar** [comment]
@sipa done

**Comment by sipa** [comment]
utACK 176e19b571f722437043d2aa80a69ae21852c70d (same tree as before)

**Pull request merged**